File: ruby/lib/coding_interview/application/usecase/market_price/update_market_price_usecase.rb
module CodingInterview
  module Application
    module Usecase
      module MarketPrice
        UpdateMarketPriceItemInput = Struct.new(:symbol, :market_price)
        UpdateMarketPriceUsecaseInput = Struct.new(:items)

        class UpdateMarketPriceUsecase
          def initialize(market_price_repository)
            @market_price_repository = market_price_repository
          end

          def run(input)
            prices = input.items.each_with_object({}) { |i, h| h[i.symbol] = i.market_price }
            @market_price_repository.update(prices)
          end
        end
      end
    end
  end
end

File: python/src/coding_interview/application/usecase/market_price/update_market_price_usecase.py
from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal

from coding_interview.application.repository.market_price_repository import MarketPriceRepository
from coding_interview.domain.stock_symbol import StockSymbol


@dataclass(frozen=True)
class UpdateMarketPriceItemInput:
    symbol: StockSymbol
    market_price: Decimal


@dataclass(frozen=True)
class UpdateMarketPriceUsecaseInput:
    items: tuple[UpdateMarketPriceItemInput, ...]


class UpdateMarketPriceUsecase:
    def __init__(self, market_price_repository: MarketPriceRepository) -> None:
        self._market_price_repository = market_price_repository

    def run(self, input: UpdateMarketPriceUsecaseInput) -> None:
        prices = {item.symbol: item.market_price for item in input.items}
        self._market_price_repository.update(prices)
